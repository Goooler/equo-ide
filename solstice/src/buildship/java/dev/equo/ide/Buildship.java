/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import org.eclipse.buildship.core.GradleBuild;
import org.eclipse.buildship.core.GradleCore;
import org.eclipse.buildship.core.GradleDistribution;
import org.eclipse.buildship.ui.internal.util.gradle.GradleDistributionViewModel;
import org.eclipse.buildship.ui.internal.wizard.project.ProjectImportConfiguration;
import org.osgi.framework.BundleContext;

class Buildship implements IdeHookInstantiated {
	IdeHookBuildship data;

	/**
	 * Logic comes from {@link org.eclipse.buildship.ui.internal.wizard.project.ProjectImportWizard}
	 */
	Buildship(IdeHookBuildship data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void initialize() {
		if (isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		var configuration = new ProjectImportConfiguration();
		configuration.setProjectDir(data.rootDir);
		configuration.setOverwriteWorkspaceSettings(false);
		configuration.setDistribution(GradleDistributionViewModel.from(GradleDistribution.fromBuild()));
		configuration.setOfflineMode(data.isOffline);

		var internalBuildConfiguration = configuration.toInternalBuildConfiguration();
		boolean showExecutionsView =
				internalBuildConfiguration.getWorkspaceConfiguration().isShowExecutionsView();
		var buildConfiguration = configuration.toBuildConfiguration();
		GradleBuild gradleBuild = GradleCore.getWorkspace().createBuild(buildConfiguration);

		// TODO: very incomplete.
	}
}
