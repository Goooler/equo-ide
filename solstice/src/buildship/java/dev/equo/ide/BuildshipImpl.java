/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import com.google.common.collect.ImmutableList;
import java.lang.reflect.InvocationTargetException;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import org.eclipse.buildship.core.GradleCore;
import org.eclipse.buildship.core.GradleDistribution;
import org.eclipse.buildship.core.SynchronizationResult;
import org.eclipse.buildship.core.internal.DefaultGradleBuild;
import org.eclipse.buildship.core.internal.workspace.NewProjectHandler;
import org.eclipse.buildship.ui.internal.util.gradle.GradleDistributionViewModel;
import org.eclipse.buildship.ui.internal.util.workbench.WorkbenchUtils;
import org.eclipse.buildship.ui.internal.util.workbench.WorkingSetUtils;
import org.eclipse.buildship.ui.internal.wizard.project.ProjectImportConfiguration;
import org.eclipse.core.resources.IProject;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.swt.widgets.Display;
import org.eclipse.ui.IWorkingSet;
import org.eclipse.ui.IWorkingSetManager;
import org.eclipse.ui.PlatformUI;
import org.eclipse.ui.internal.Workbench;
import org.gradle.tooling.GradleConnector;
import org.osgi.framework.BundleContext;

class BuildshipImpl implements IdeHookInstantiated {
	IdeHookBuildship data;

	/**
	 * Logic comes from {@link org.eclipse.buildship.ui.internal.wizard.project.ProjectImportWizard}
	 */
	BuildshipImpl(IdeHookBuildship data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void postStartup() throws Exception {
		if (!isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		Workbench.getInstance().getActiveWorkbenchWindow().run(true, true, this::doImport);
	}

	private void doImport(IProgressMonitor monitor) throws InvocationTargetException {
		var importCfg = new ProjectImportConfiguration();
		importCfg.setProjectDir(data.rootDir);
		importCfg.setOfflineMode(data.isOffline);

		var distribution = GradleDistribution.fromBuild();
		importCfg.setDistribution(GradleDistributionViewModel.from(distribution));

		importCfg.setAutoSync(true);
		importCfg.setApplyWorkingSets(true);
		importCfg.setBuildScansEnabled(false);
		importCfg.setShowConsoleView(true);
		importCfg.setShowExecutionsView(true);
		importCfg.setOverwriteWorkspaceSettings(true);
		importCfg.setWorkingSets(Collections.emptyList());

		var buildCfg = importCfg.toBuildConfiguration();
		var gradleBuild = GradleCore.getWorkspace().createBuild(buildCfg);
		var workingSetsAddingNewProjectHandler =
				new ImportWizardNewProjectHandler(
						NewProjectHandler.IMPORT_AND_MERGE,
						importCfg,
						importCfg.getShowExecutionsView().getValue());
		SynchronizationResult result =
				((DefaultGradleBuild) gradleBuild)
						.synchronize(
								workingSetsAddingNewProjectHandler,
								GradleConnector.newCancellationTokenSource(),
								monitor);
		if (!result.getStatus().isOK()) {
			throw new InvocationTargetException(new CoreException(result.getStatus()));
		}
	}

	private static final class ImportWizardNewProjectHandler implements NewProjectHandler {
		private final ProjectImportConfiguration configuration;
		private final NewProjectHandler importedBuildDelegate;
		private final boolean showExecutionsView;
		private volatile boolean gradleViewsVisible;

		private ImportWizardNewProjectHandler(
				NewProjectHandler delegate,
				ProjectImportConfiguration configuration,
				boolean showExecutionsView) {
			this.importedBuildDelegate = delegate;
			this.configuration = configuration;
			this.showExecutionsView = showExecutionsView;
		}

		public boolean shouldImportNewProjects() {
			return this.importedBuildDelegate.shouldImportNewProjects();
		}

		public void afterProjectImported(IProject project) {
			this.importedBuildDelegate.afterProjectImported(project);
			this.addWorkingSets(project);
			this.ensureGradleViewsAreVisible();
		}

		private void addWorkingSets(IProject project) {
			List<String> workingSetNames =
					configuration.getApplyWorkingSets().getValue()
							? ImmutableList.copyOf(
									(Collection<String>) this.configuration.getWorkingSets().getValue())
							: ImmutableList.of();
			IWorkingSetManager workingSetManager = PlatformUI.getWorkbench().getWorkingSetManager();
			IWorkingSet[] workingSets = WorkingSetUtils.toWorkingSets(workingSetNames);
			workingSetManager.addToWorkingSets(project, workingSets);
		}

		private void ensureGradleViewsAreVisible() {
			Display.getDefault()
					.asyncExec(
							() -> {
								WorkbenchUtils.showView(
										"org.eclipse.buildship.ui.views.taskview", (String) null, 1);
								WorkbenchUtils.showView(
										"org.eclipse.buildship.ui.views.executionview", (String) null, 2);
							});
		}
	}
}
