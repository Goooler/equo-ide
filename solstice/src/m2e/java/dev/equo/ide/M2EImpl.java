/*******************************************************************************
 * Copyright (c) 2023 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.io.File;
import java.lang.reflect.InvocationTargetException;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.m2e.core.MavenPlugin;
import org.eclipse.m2e.core.embedder.MavenModelManager;
import org.eclipse.m2e.core.project.LocalProjectScanner;
import org.eclipse.m2e.core.project.MavenProjectInfo;
import org.eclipse.m2e.core.project.ProjectImportConfiguration;
import org.eclipse.m2e.core.ui.internal.wizards.ImportMavenProjectsJob;
import org.eclipse.ui.IWorkingSet;
import org.eclipse.ui.internal.Workbench;
import org.osgi.framework.BundleContext;

class M2EImpl implements IdeHookInstantiated {
	IdeHookM2E data;

	M2EImpl(IdeHookM2E data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void postStartup() throws Exception {
		if (!isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		Workbench.getInstance().getActiveWorkbenchWindow().run(true, true, this::doImport);
	}

	private LocalProjectScanner instantiateScanner() {
		boolean basedirRemameRequired = false;
		var list = List.of(data.rootDir.getAbsolutePath());
		try {
			try {
				// 2.1.2 (Java 17+)
				var c =
						LocalProjectScanner.class.getConstructor(
								List.class, boolean.class, MavenModelManager.class);
				return c.newInstance(list, basedirRemameRequired, MavenPlugin.getMavenModelManager());
			} catch (NoSuchMethodException e) {
				// 1.20.1 (Java 11+)
				var c =
						LocalProjectScanner.class.getConstructor(
								File.class, List.class, boolean.class, MavenModelManager.class);
				return c.newInstance(
						data.rootDir, list, basedirRemameRequired, MavenPlugin.getMavenModelManager());
			}
		} catch (InvocationTargetException
				| InstantiationException
				| IllegalAccessException
				| NoSuchMethodException e) {
			throw new RuntimeException(e);
		}
	}

	private void doImport(IProgressMonitor monitor)
			throws InvocationTargetException, InterruptedException {
		var discoverMonitor = monitor.slice(20);
		var importMonitor = monitor.slice(80);

		var scanner = instantiateScanner();
		scanner.run(discoverMonitor);

		var importConfiguration = new ProjectImportConfiguration();
		var workingSets = List.<IWorkingSet>of();

		var projects = new HashSet<MavenProjectInfo>();
		for (var p : scanner.getProjects()) {
			addAll(p, projects);
		}
		var job = new ImportMavenProjectsJob(projects, workingSets, importConfiguration);
		var status = job.runInWorkspace(importMonitor);
		if (!status.isOK()) {
			throw new InvocationTargetException(new CoreException(status));
		}
		monitor.done();
	}

	private void addAll(MavenProjectInfo proj, Collection<MavenProjectInfo> projects) {
		projects.add(proj);
		for (var child : proj.getProjects()) {
			addAll(child, projects);
		}
	}
}
