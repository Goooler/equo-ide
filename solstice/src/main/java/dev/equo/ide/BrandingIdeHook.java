/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.util.Arrays;
import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class BrandingIdeHook implements IdeHook {
	@Override
	public IdeHookInstantiated instantiate() {
		return new Instantiated();
	}

	class Instantiated implements IdeHookInstantiated {
		private Shell splash;

		@Override
		public void afterDisplay(org.eclipse.swt.widgets.Display display) {
			var cursor = display.getCursorLocation();
			var monitors = display.getMonitors();
			var bestMonitor =
					Arrays.stream(monitors)
							.filter(monitor -> monitor.getBounds().contains(cursor))
							.findAny()
							.orElse(monitors[0])
							.getBounds();

			int imgX = 1000;
			int imgY = 1000;
			splash = new Shell(display, SWT.NO_TRIM);
			splash.setText("Branding");
			splash.setBounds(
					bestMonitor.x + (bestMonitor.width - imgX) / 2,
					bestMonitor.y + (bestMonitor.height - imgY) / 2,
					imgX,
					imgY);
			splash.setBackground(display.getSystemColor(SWT.COLOR_GREEN));
			splash.open();

			Display.setAppName("Branding");
			while (display.readAndDispatch())
				// pump the event loop enough to show the branding
				;
		}

		@Override
		public void postStartup() {
			splash.dispose();
			splash = null;

			Shell[] shells = Display.getCurrent().getShells();
			for (var shell : shells) {
				shell.setText("Branding");
			}
		}
	}
}
