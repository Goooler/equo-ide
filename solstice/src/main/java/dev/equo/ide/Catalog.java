/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;

public class Catalog implements Comparable<Catalog> {
	private static String V = "${VERSION}";
	public static final Catalog PLATFORM =
			new Catalog(
					"platform",
					"https://download.eclipse.org/eclipse/updates/" + V,
					"4.26",
					"org.eclipse.platform.ide.categoryIU");
	public static final Catalog JDT =
			new Catalog("jdt", PLATFORM, "org.eclipse.releng.java.languages.categoryIU", PLATFORM);
	public static final Catalog GRADLE_BUILDSHIP =
			new Catalog(
					"gradleBuildship",
					"https://download.eclipse.org/buildship/updates/e423/releases/3.x/" + V,
					"3.1.6.v20220511-1359",
					"org.eclipse.buildship.feature.group",
					JDT);

	private final String name;
	private final String p2urlTemplate;
	private final String latestVersion;
	private final String toInstall;
	private final List<Catalog> requires;

	Catalog(String name, Catalog copyFrom, String toInstall, Catalog... requires) {
		this(name, copyFrom.p2urlTemplate, copyFrom.latestVersion, toInstall, requires);
	}

	Catalog(
			String name,
			String p2urlTemplate,
			String latestVersion,
			String toInstall,
			Catalog... requires) {
		this.name = name;
		if (p2urlTemplate.endsWith("/")) {
			this.p2urlTemplate = p2urlTemplate;
		} else {
			this.p2urlTemplate = p2urlTemplate + "/";
		}
		this.latestVersion = latestVersion;
		this.toInstall = toInstall;
		this.requires = Arrays.asList(requires);
	}

	public String getP2UrlTemplate() {
		return p2urlTemplate;
	}

	public String getName() {
		return name;
	}

	public List<Catalog> getRequires() {
		return requires;
	}

	public String getUrlForOverride(@Nullable String override) {
		if (override == null) {
			return p2urlTemplate.replace(V, latestVersion);
		} else if (isUrl(override)) {
			return override;
		} else {
			return p2urlTemplate.replace(V, override);
		}
	}

	public List<String> getTargetsFor(@Nullable String override) {
		return Collections.singletonList(toInstall);
	}

	public static boolean isUrl(String maybeUrl) {
		return maybeUrl.startsWith("https:") || maybeUrl.startsWith("http:");
	}

	@Override
	public int compareTo(Catalog o) {
		return name.compareTo(o.name);
	}

	@Override
	public int hashCode() {
		return name.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Catalog) {
			return ((Catalog) obj).name.equals(name);
		} else {
			return false;
		}
	}
}
