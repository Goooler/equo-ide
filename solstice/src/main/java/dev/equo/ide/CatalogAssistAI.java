/*******************************************************************************
 * Copyright (c) 2023 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.util.List;

public class CatalogAssistAI extends Catalog {
	CatalogAssistAI() {
		super(
				"assistAI",
				"https://eclipse-chatgpt-plugin.lm.r.appspot.com/",
				jre17(""),
				List.of("com.github.gradusnikov.eclipse.assistai"),
				PLATFORM,
				ORBIT,
				EGIT);
	}

	private static final String SETTINGS_FILE =
			"instance/.metadata/.plugins/org.eclipse.core.runtime/.settings/com.github.gradusnikov.eclipse.plugin.assistai.main.prefs";

	public void apiKey(WorkspaceInit workspace, String apiKey) {
		workspace.setProperty(SETTINGS_FILE, "OpenAIAPIKey", apiKey);
	}

	public void modelName(WorkspaceInit workspace, String modelName) {
		workspace.setProperty(SETTINGS_FILE, "OpenAIModelName", modelName);
	}
}
