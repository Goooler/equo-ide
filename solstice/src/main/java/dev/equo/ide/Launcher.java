/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.io.BufferedOutputStream;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.function.Consumer;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;
import javax.annotation.Nullable;

/**
 * Thanks to Thipor Kong for his workaround for Gradle's windows problems.
 *
 * <p>https://discuss.gradle.org/t/javaexec-fails-for-long-classpaths-on-windows/15266
 */
public class Launcher {
	public static int launchJavaBlocking(
			boolean blocking,
			List<File> cp,
			List<String> vmArgs,
			String mainClass,
			@Nullable Consumer<Process> monitorProcess,
			String... args)
			throws IOException, InterruptedException {
		String javaHome = System.getProperty("java.home");
		String javaBin = javaHome + File.separator + "bin" + File.separator + "java";
		String javaCmd;
		if (new File(javaBin).exists()) {
			javaCmd = javaBin;
		} else {
			javaCmd = "java";
		}

		File classpathJar = Launcher.toJarWithClasspath(cp);
		classpathJar.deleteOnExit();

		List<String> command = new ArrayList<>();
		command.add(javaCmd);
		command.addAll(vmArgs);
		command.add("-classpath");
		command.add(classpathJar.getAbsolutePath());
		command.add(mainClass);
		command.addAll(Arrays.asList(args));

		if (blocking) {
			return launchAndInheritIO(null, command, monitorProcess);
		} else {
			ScriptExec.script(ScriptExec.quoteAll(command)).execSeparate(monitorProcess);
			return 0;
		}
	}

	public static int launchAndInheritIO(File cwd, List<String> args)
			throws IOException, InterruptedException {
		return launchAndInheritIO(cwd, args, null);
	}

	public static int launchAndInheritIO(
			File cwd, List<String> args, @Nullable Consumer<Process> monitorProcess)
			throws IOException, InterruptedException {
		var builder = new ProcessBuilder(args);
		if (cwd != null) {
			builder.directory(cwd);
		}
		var process = builder.start();
		var outPumper = new StreamPumper(process, process.getInputStream(), System.out);
		var errPumper = new StreamPumper(process, process.getErrorStream(), System.err);
		if (monitorProcess != null) {
			new Thread(() -> monitorProcess.accept(process)).start();
		}
		int exitCode = process.waitFor();
		process.getOutputStream().flush();
		outPumper.join();
		errPumper.join();
		return exitCode;
	}

	static class StreamPumper extends Thread {
		private final Process process;
		private final InputStream in;
		private final PrintStream out;

		private StreamPumper(Process process, InputStream in, PrintStream out) {
			this.process = process;
			this.in = in;
			this.out = out;
			start();
		}

		@Override
		public void run() {
			byte[] buf = new byte[1024];
			try {
				int numRead;
				while ((numRead = in.read(buf)) != -1) {
					out.write(buf, 0, numRead);
					out.flush();
				}
			} catch (IOException e) {
				if (process.isAlive()) {
					// only report the exception if the process is alive
					throw new RuntimeException(e);
				}
			}
		}
	}

	private static final String LONG_CLASSPATH_JAR_PREFIX = "long-classpath";

	/** Creates a jar with a Class-Path entry to workaround the windows classpath limitation. */
	private static File toJarWithClasspath(List<File> files) throws IOException {
		File jarFile = File.createTempFile(LONG_CLASSPATH_JAR_PREFIX, ".jar");
		try (ZipOutputStream zip =
				new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(jarFile)))) {
			zip.putNextEntry(new ZipEntry("META-INF/MANIFEST.MF"));
			try (PrintWriter pw =
					new PrintWriter(
							new BufferedWriter(new OutputStreamWriter(zip, StandardCharsets.UTF_8)))) {
				pw.println("Manifest-Version: 1.0");
				StringBuilder bufferClassPath = new StringBuilder("Class-Path: ");
				for (File file : files) {
					if (bufferClassPath.length() != 0) {
						bufferClassPath.append(' ');
					}
					bufferClassPath.append(file.toURI());
				}
				pw.println(
						String.join("\n ", bufferClassPath.toString().split(MATCH_CHUNKS_OF_70_CHARACTERS)));
			}
		}
		return jarFile;
	}

	public static ArrayList<File> copyAndSortClasspath(Iterable<File> files) {
		var copy = new ArrayList<File>();
		List<File> addToEnd = new ArrayList<>();
		for (File f : files) {
			if (f.getName().startsWith("slf4j-nop")) {
				continue;
			}
			if (f.getName().startsWith("org.apache.jasper.glassfish")
					|| f.getName().startsWith("biz.aQute.bndlib")) {
				addToEnd.add(f);
				continue;
			}
			copy.add(f);
		}
		copy.sort(Comparator.comparing(File::getName).reversed());
		copy.addAll(addToEnd);
		return copy;
	}

	private static final String MATCH_CHUNKS_OF_70_CHARACTERS = "(?<=\\G.{70})";
}
