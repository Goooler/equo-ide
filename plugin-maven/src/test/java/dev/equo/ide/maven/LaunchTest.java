/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide.maven;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import dev.equo.ide.Launcher;
import java.io.IOException;
import java.util.Arrays;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({SnapshotExtension.class})
public class LaunchTest extends MavenHarness {
	@Test
	public void help(Expect expect) throws IOException, InterruptedException {
		setPom("");
		mvnw("help:describe -Dcmd=equo-ide:launch -Ddetail")
				.snapshotBetween("Mojo: 'equo-ide:launch'", "[INFO] BUILD SUCCESS", expect);
	}

	@Test
	public void integrationTestAtomos() throws IOException, InterruptedException {
		integrationTestUseAtomos(true);
	}

	@Test
	public void integrationTestSolstice() throws IOException, InterruptedException {
		integrationTestUseAtomos(false);
	}

	private void integrationTestUseAtomos(boolean useAtomos)
			throws IOException, InterruptedException {
		setPom(
				"<p2repos>\n"
						+ "  <p2repo>https://download.eclipse.org/eclipse/updates/4.26/</p2repo>\n"
						+ "</p2repos>\n"
						+ "<installs>\n"
						+ "  <install>org.eclipse.swt</install>\n"
						+ "</installs>");
		var output = mvnw("equo-ide:launch -DinitOnly -DuseAtomos=" + useAtomos);
		output.raw().matches("(?s)(.*)Loaded (\\d+) bundles(.*)");
		output.raw().contains(useAtomos ? "using Atomos" : "not using Atomos");
	}

	@Disabled
	@Test
	public void integrationTestReal() throws IOException, InterruptedException {
		setPom("");
		Launcher.launchAndInheritIO(rootFolder(), Arrays.asList("mvn", "equo-ide:launch", "-q"));
	}
}
