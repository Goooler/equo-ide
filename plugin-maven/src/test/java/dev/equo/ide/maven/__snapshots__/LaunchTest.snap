dev.equo.ide.maven.LaunchTest.help=[
equo-ide:launch
  Description: Launches an Eclipse-based IDE for this project.
  Implementation: dev.equo.ide.maven.LaunchMojo
  Language: java

  Available parameters:

    branding
      (no description available)

    cdt
      (no description available)

    clean (Default: false)
      User property: clean
      Wipes all IDE settings and state before rebuilding and launching.

    debugClasspath (Default: disabled)
      User property: debugClasspath
      Dumps the classpath (in order) without starting the application.

    debugIde (Default: false)
      User property: debugIde
      Blocks IDE startup to help you attach a debugger.

    equoChromium (Default: false)
      User property: equoChromium
      Replaces the standard SWT browser with Equo Chromium.

    filters
      (no description available)

    gradleBuildship
      (no description available)

    groovy
      (no description available)

    initOnly (Default: false)
      User property: initOnly
      Initializes the runtime to check for errors then exits.

    installs
      (no description available)

    jdt
      (no description available)

    kotlin
      (no description available)

    m2e
      (no description available)

    p2repos
      (no description available)

    pde
      (no description available)

    platform
      (no description available)

    rust
      (no description available)

    showConsole (Default: false)
      User property: showConsole
      Adds a visible console to the launched application.

    tmTerminal
      (no description available)

    useAtomos (Default: false)
      User property: useAtomos
      Determines whether to use Solstice's built-in OSGi runtime or instead
      Atomos+Equinox.

    welcome
      (no description available)
.
.
[INFO] ------------------------------------------------------------------------
]